lib/ansible/modules/cloud/amazon/storage_gateway.py

```
@@ -0,0 +1,241 @@
+    default: 'STORED'
+    choices: ['STORED', 'CACHED', 'VTL', 'FILE_S3']
+  tape_drive_type:
+    description:
+    - The value that indicates the type of tape drive to use for tape gateway.
+    choices: ['IBM-ULT3580-TD5']
+  medium_changer_type:
+    description:
+    - The value that indicates the type of medium changer to use for tape gateway.
+    choices: ['STK-L700', 'AWS-Gateway-VTL']
+extends_documentation_fragment:
+  - aws
+  - ec2
+'''
+
+EXAMPLES = r'''
+  - name: Activate file gateway
+    storage_gateway:
+      name: "example-file-gateway"
+      state: present
+      activation_key: "{{ activation_code.stdout }}"
+      timezone: 'GMT-6:00'
+      gateway_region: "{{ aws_region }}"
+      gateway_type: 'FILE_S3'
+'''
+
+RETURN = r'''
+gateway_arn:
+    description: The ARN of the gateway you just created or updated.
+    returned: always
+    type: string
+'''
+
+import time
+import ast
+
+try:
+    import botocore
+    from botocore.exceptions import BotoCoreError, ClientError
+except ImportError:
+    pass  # handled by AnsibleAWSModule
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, get_aws_connection_info, AWSRetry
+from ansible.module_utils.ec2 import camel_dict_to_snake_dict, boto3_tag_list_to_ansible_dict
+
+
+def gateway_exists(client, module, params, result):
+    try:
+        gateway_list = client.list_gateways()
+        for i in gateway_list['Gateways']:
+            if i['GatewayName'] == params['GatewayName']:
+                disks_response = client.list_local_disks(
+                    GatewayARN=i['GatewayARN']
+                )
+                result['GatewayARN'] = i['GatewayARN']
+                result['Disks'] = disks_response['Disks']
+                return True
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError):
+        return False
```

It looks like list_gateways() returns an empty list rather than failing if no gateways exist, so I think fail_json_aws should probably be used here. Right now valid ClientError and BotoCoreError exceptions are masked. If there's a specific boto3 exception that needs to be handled, use `except is_boto3_error_code(...` right before line 109 .